File: appMain/src/main/java/com/cartlc/tracker/app/TBApplication.java
/**
 * Copyright 2018, FleetTLC. All rights reserved
 */
package com.cartlc.tracker.app;

import android.Manifest;
import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.support.multidex.MultiDex;
import android.support.v4.content.FileProvider;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import com.cartlc.tracker.BuildConfig;
import com.cartlc.tracker.R;
import com.cartlc.tracker.data.DatabaseManager;
import com.cartlc.tracker.etc.CheckError;
import com.cartlc.tracker.etc.PrefHelper;
import com.cartlc.tracker.etc.BootstrapData;
import com.cartlc.tracker.data.TableZipCode;
import com.cartlc.tracker.data.DataZipCode;
import com.cartlc.tracker.event.EventError;
import com.cartlc.tracker.server.AmazonHelper;
import com.cartlc.tracker.server.DCService;
import com.cartlc.tracker.server.ServerHelper;
import com.cartlc.tracker.util.LocationHelper;
import com.cartlc.tracker.util.PermissionHelper.PermissionRequest;
import com.cartlc.tracker.util.PermissionHelper.PermissionListener;

import de.greenrobot.event.EventBus;
import timber.log.Timber;

import com.cartlc.tracker.util.PermissionHelper;
import com.squareup.leakcanary.LeakCanary;

import java.io.File;

public class TBApplication extends Application {

    static final Boolean OVERRIDE_IS_DEVELOPMENT_SERVER = null;

    public static boolean IsDevelopmentServer() {
        if (OVERRIDE_IS_DEVELOPMENT_SERVER != null) {
            return OVERRIDE_IS_DEVELOPMENT_SERVER;
        }
        return BuildConfig.DEBUG;
    }

    static final Boolean DEBUG_TREE = false;
    static final Boolean LEAK_CANARY = false;

    public static final Boolean REPORT_LOCATION = true; // BuildConfig.DEBUG;
    public static final boolean LOCATION_ENABLE = true;

    public static final String OTHER = "Other";

    CrashReportingTree mCrashTree = new CrashReportingTree();

    static final PermissionRequest[] PERMISSIONS = new PermissionRequest[]{
            new PermissionRequest(Manifest.permission.WRITE_EXTERNAL_STORAGE, R.string.perm_read_external_storage),
            new PermissionRequest(Manifest.permission.READ_EXTERNAL_STORAGE, R.string.perm_write_external_storage),
            new PermissionRequest(Manifest.permission.ACCESS_FINE_LOCATION, R.string.perm_location),
    };

    public TBApplication() {
        super();
    }

    @Override
    public void onCreate() {
        super.onCreate();

        if (LEAK_CANARY) {
            if (LeakCanary.isInAnalyzerProcess(this)) {
                // This process is dedicated to LeakCanary for heap analysis.
                // You should not init your app in this process.
                return;
            }
            LeakCanary.install(this);
        }
        if (IsDevelopmentServer() && DEBUG_TREE) {
            Timber.plant(new Timber.DebugTree());
        } else {
            Timber.plant(mCrashTree);
        }
        DatabaseManager.Init(this);
        PrefHelper.Init(this);
        ServerHelper.Init(this);
        PermissionHelper.Init();
        AmazonHelper.Init(this);
        BootstrapData.Init();
        CheckError.Init();

        PrefHelper.getInstance().detectSpecialUpdateCheck();
        LocationHelper.Init(this);
    }

    public void ping() {
        if (ServerHelper.getInstance().hasConnection()) {
            startService(new Intent(this, DCService.class));
        }
    }

    public void requestZipCode(String zipCode) {
        DataZipCode data = TableZipCode.getInstance().query(zipCode);
        if (data != null) {
            data.check();
            EventBus.getDefault().post(data);
        } else if (ServerHelper.getInstance().hasConnection()) {
            Intent intent = new Intent(this, DCService.class);
            intent.setAction(DCService.ACTION_ZIP_CODE);
            intent.putExtra(DCService.DATA_ZIP_CODE, zipCode);
            startService(intent);
        }
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        MultiDex.install(this);
    }

    public static Uri getUri(Context ctx, File file) {
        return FileProvider.getUriForFile(ctx, "com.cartcl.tracker.fileprovider", file);
    }

    public static void hideKeyboard(Context ctx, View v) {
        InputMethodManager imm = (InputMethodManager) ctx.getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
    }

    public void setUncaughtExceptionHandler(final Activity act) {
//        Thread.setDefaultUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
//            @Override
//            public void uncaughtException(Thread t, Throwable e) {
//                mCrashTree.log(Log.ERROR, "CarTLCarTLC", e.getMessage(), null);
//                act.finish();
//            }
//        });
    }

    public void checkPermissions(Activity act, PermissionListener listener) {
        PermissionHelper.getInstance().checkPermissions(act, PERMISSIONS, listener);
    }

    public static String ReportError(Exception ex, Class claz, String function, String type) {
        return ReportError(ex.getMessage(), claz, function, type);
    }

    public static String ReportError(String msg, Class claz, String function, String type) {
        StringBuilder sbuf = new StringBuilder();
        sbuf.append("Class:");
        sbuf.append(claz.getSimpleName());
        sbuf.append(".");
        sbuf.append(function);
        sbuf.append(" ");
        sbuf.append(type);
        sbuf.append(": ");
        sbuf.append(msg);
        Timber.e(sbuf.toString());
        return sbuf.toString();
    }

    public String getVersion() throws PackageManager.NameNotFoundException {
        StringBuilder sbuf = new StringBuilder();
        String version = getPackageManager().getPackageInfo(getPackageName(), 0).versionName;
        sbuf.append("v");
        sbuf.append(version);
        if (PrefHelper.getInstance().isDevelopment()) {
            sbuf.append("d");
        }
        return sbuf.toString();
    }

    public static void ReportServerError(Exception ex, Class claz, String function, String type) {
        String msg = ReportError(ex, claz, function, type);
        ShowError(msg);
    }

    public static void ShowError(String msg) {
        EventBus.getDefault().post(new EventError(msg));
    }
}

File: appMain/src/main/java/com/cartlc/tracker/ui/app/TBApplication.kt
/**
 * Copyright 2018, FleetTLC. All rights reserved
 */
package com.cartlc.tracker.ui.app

import android.Manifest
import android.app.Activity
import android.app.Application
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import androidx.core.content.FileProvider
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.annotation.VisibleForTesting
import androidx.multidex.MultiDex

import com.cartlc.tracker.BuildConfig
import com.cartlc.tracker.R
import com.cartlc.tracker.model.*
import com.cartlc.tracker.ui.util.CheckError
import com.cartlc.tracker.model.pref.PrefHelper
import com.cartlc.tracker.model.event.EventError
import com.cartlc.tracker.model.server.AmazonHelper
import com.cartlc.tracker.model.server.DCService
import com.cartlc.tracker.model.server.ServerHelper
import com.cartlc.tracker.model.sql.DatabaseManager
import com.cartlc.tracker.model.table.DatabaseTable
import com.cartlc.tracker.ui.app.dependencyinjection.ComponentRoot
import com.cartlc.tracker.ui.util.helper.LocationHelper
import com.cartlc.tracker.ui.util.helper.PermissionHelper.PermissionRequest
import com.cartlc.tracker.ui.util.helper.PermissionHelper.PermissionListener

import com.cartlc.tracker.ui.util.helper.PermissionHelper
import com.cartlc.tracker.viewmodel.vehicle.DaggerVehicleViewModelComponent
import com.cartlc.tracker.viewmodel.vehicle.VehicleViewModel
import com.cartlc.tracker.viewmodel.vehicle.VehicleViewModelComponent
import com.cartlc.tracker.viewmodel.vehicle.VehicleViewModelModule
import com.squareup.leakcanary.LeakCanary

import org.greenrobot.eventbus.EventBus
import timber.log.Timber
import java.io.File

class TBApplication : Application() {

    companion object {

        private val OVERRIDE_IS_DEVELOPMENT_SERVER: Boolean? = null

        @VisibleForTesting
        var DEBUG_TREE = false

        private const val LEAK_CANARY = false

        const val REPORT_LOCATION = false
        const val OTHER = "Other"

        private val PERMISSIONS = arrayOf(
                PermissionRequest(Manifest.permission.WRITE_EXTERNAL_STORAGE,
                        R.string.perm_read_external_storage),
                PermissionRequest(Manifest.permission.READ_EXTERNAL_STORAGE,
                        R.string.perm_write_external_storage),
                PermissionRequest(Manifest.permission.ACCESS_FINE_LOCATION,
                        R.string.perm_location))

        fun IsDevelopmentServer(): Boolean {
            return OVERRIDE_IS_DEVELOPMENT_SERVER ?: BuildConfig.DEBUG
        }

        fun getUri(ctx: Context, file: File): Uri {
            return FileProvider.getUriForFile(ctx, "com.cartcl.tracker.fileprovider", file)
        }

        fun hideKeyboard(ctx: Context, v: View) {
            val imm = ctx.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(v.windowToken, 0)
        }

        fun ReportError(ex: Exception, claz: Class<*>, function: String, type: String): String {
            return ReportError(ex.message
                    ?: "unknown", claz, function, type)
        }

        fun ReportError(msg: String, claz: Class<*>, function: String, type: String): String {
            val sbuf = StringBuilder()
            sbuf.append("Class:")
            sbuf.append(claz.simpleName)
            sbuf.append(".")
            sbuf.append(function)
            sbuf.append(" ")
            sbuf.append(type)
            sbuf.append(": ")
            sbuf.append(msg)
            Timber.e(sbuf.toString())
            return sbuf.toString()
        }

        fun ReportServerError(ex: Exception, claz: Class<*>, function: String, type: String) {
            val msg = ReportError(ex, claz, function, type)
            ShowError(msg)
        }

        fun ShowError(msg: String) {
            EventBus.getDefault().post(EventError(msg))
        }
    }

    val versionedTitle: String
        get() {
            val sbuf = StringBuilder()
            sbuf.append(getString(R.string.app_name))
            sbuf.append(" - ")
            try {
                sbuf.append(version)
            } catch (ex: Exception) {
                ReportError(ex, TBApplication::class.java, "versionedTitle", "main")
            }
            return sbuf.toString()
        }

//    lateinit var carRepoComponent: CarRepositoryComponent

    private lateinit var carRepo: CarRepository
    private lateinit var prefHelper: PrefHelper
    private lateinit var dm: DatabaseManager

    val repo: CarRepository
        get() = carRepo

    lateinit var componentRoot: ComponentRoot
    lateinit var vehicleComponent: VehicleViewModelComponent
    lateinit var amazonHelper: AmazonHelper

    private lateinit var vehicleViewModel: VehicleViewModel
    private lateinit var vehicleRepository: VehicleRepository

    val db: DatabaseTable
        get() = dm

    val version: String
        @Throws(PackageManager.NameNotFoundException::class)
        get() {
            val sbuf = StringBuilder()
            val version = packageManager.getPackageInfo(packageName, 0).versionName
            sbuf.append("v")
            sbuf.append(version)
            if (prefHelper.isDevelopment) {
                sbuf.append("d")
            }
            return sbuf.toString()
        }

    override fun onCreate() {
        super.onCreate()

        componentRoot = ComponentRoot(this)
        dm = DatabaseManager(this)
        prefHelper = PrefHelper(this, dm)
        carRepo = CarRepository(
                dm,
                prefHelper,
                componentRoot.flowUseCase
        )
        carRepo.computeCurStage()

        vehicleRepository = VehicleRepository(this, dm, prefHelper)
        vehicleViewModel = VehicleViewModel(vehicleRepository)
        vehicleComponent = DaggerVehicleViewModelComponent.builder()
                .vehicleViewModelModule(VehicleViewModelModule(vehicleViewModel))
                .build()

        if (LEAK_CANARY) {
            if (LeakCanary.isInAnalyzerProcess(this)) {
                // This process is dedicated to LeakCanary for heap analysis.
                // You should not init your app in this process.
                return
            }
            LeakCanary.install(this)
        }
        if (IsDevelopmentServer() && DEBUG_TREE) {
            Timber.plant(Timber.DebugTree())
        } else {
            Timber.plant(CrashReportingTree(db))
        }
        ServerHelper.Init(this)
        amazonHelper = AmazonHelper(db, prefHelper, componentRoot.eventController)
        PermissionHelper.Init()
        CheckError.Init()
        LocationHelper.Init(this, db)

        if (prefHelper.detectOneTimeReloadFromServerCheck()) {
            reloadFromServer()
        }
    }

    fun reloadFromServer() {
        prefHelper.reloadFromServer()
        ping()
    }

    fun ping() {
        if (ServerHelper.instance.hasConnection(this)) {
            startService(Intent(this, DCService::class.java))
        }
    }

//    fun requestZipCode(tableZipCode: String) {
//        val data = db.tableZipCode.query(tableZipCode)
//        if (data != null) {
//            data.check()
//            EventBus.getDefault().post(data)
//        } else if (ServerHelper.instance.hasConnection(this)) {
//            val intent = Intent(this, DCService::class.java)
//            intent.action = DCService.ACTION_ZIP_CODE
//            intent.putExtra(DCService.DATA_ZIP_CODE, tableZipCode)
//            startService(intent)
//        }
//    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        MultiDex.install(this)
    }

    fun checkPermissions(act: Activity, listener: PermissionListener) {
        PermissionHelper.instance.checkPermissions(act, PERMISSIONS, listener)
    }

}
